README.md

```diff
@@ -291,173 +321,204 @@ be used in global variable initializers.
-surrogates, in which case the result may be a surrogate.  If *`cursor`*
-is at the end of *`str`*, the result is -1.
+Return the codepoint at the WTF-8 *`pos`* in *`str`*, as an i32.  The
+result will be a valid codepoint, and may be an isolated surrogate.  If
+*`pos`* is at the end of *`str`*, the result is -1.
 
 ```
-(string.measure $encoding str:stringref cursor:i32 codepoints:i32)
+(string.get_wtf16 str:stringref pos:i32)
+  -> codeunit:i32
+```
+Return the WTF-16 code unit at the WTF-16 *`pos`* in *`str`*, as an i32.
+The result may be part of a surrogate pair, or may be an isolated
+surrogate; the caller is responsible for decoding to codepoints as
+needed.  If *`pos`* is at the end of *`str`*, the result is -1.
+
+```
+(string.measure_utf8 str:stringref pos:i32 codepoints:i32)
+  -> bytes:i32
+(string.measure_wtf8 str:stringref pos:i32 codepoints:i32)
   -> bytes:i32
 ```
 Measure how many bytes would be necessary to encode the contents of the
-string *`str`*, starting at cursor *`cursor`*, limited to a maximum of
-*`codepoints`* codepoints.  If the contents of the string cannot be
-encoded in the given encoding, return -1.
+string *`str`*, starting at position *`pos`*, to UTF-8 or WTF-8
+respectively, limited to *`codepoints`*.  For `string.measure_utf8`, if
+the *`codepoints`* following *`pos`* contain an isolated surrogate,
+return -1.
 
 The maximum number of bytes returned by `string.measure` is
 2<sup>31</sup>-1.  If an encoding would require more bytes, it is as if
 the contents can't be encoded at all (the return value is -1).
 
-(Note that the limit on cursor values constrains strings to have a
+(Note that the limit on position values constrains strings to have a
 maximum of 2<sup>31</sup>–1 codepoints, but it may be that such a string
 could exceed even 2<sup>32</sup> bytes in a given encoding; e.g. a
 string of 2<sup>31</sup>-1 `\uFFFF` codepoints would require
 2<sup>32</sup>+2<sup>31</sup>-3 bytes when encoded as UTF-8.)
 
+There is no `string.measure_wtf16`; `string.end_wtf16` provides the
+total code unit length, and any position slice in that range is valid
+and has a well-defined mapping to bytes.
+
 ```
-(string.encode $encoding $memory str:stringref cursor:i32 ptr:address bytes:i32)
-  -> cursor:i32, codepoints:i32, bytes:i32
+(string.encode_utf8 str:stringref pos:i32 bytes:i32)
+  -> bytes:i32
+(string.encode_wtf8 str:stringref pos:i32 bytes:i32)
+  -> bytes:i32
 ```
-Encode the contents of the string *`str`*, starting at the cursor
-*`cursor`*, to memory at *`ptr`*, limited to *`bytes`* bytes.  Return
-three values: a string cursor pointing to the remaining contents of the
-string, the number of codepoints written, and the number of bytes
+
+Encode the contents of the string *`str`* as UTF-8 or WTF-8,
+respectively, starting at the WTF-8 byte offset *`pos`*, to memory at
+*`ptr`*, limited to *`bytes`* bytes.  Return the number of bytes
 written.  Note that no `NUL` terminator is ever written.  If any
```

Comment:
Would it make sense to limit the maximum output to 2^31-2 (instead of ...-1) so that callers can add their own NUL byte (if desired) and still remain in int32 range for the overall length?